docs/design/generics/terminology.md

```diff
@@ -250,21 +265,111 @@ fail.
+called _constant binding patterns_, and correspond to those binding patterns
+that use `:!`.
+
+The name being declared, which is the identifier to the left of the `:` or `:!`,
+is called a _binding_, or more specifically a _runtime binding_, _constant
+binding_, _symbolic binding_, or _template binding_. The expression to the right
+defining the type of the binding pattern is called the _binding type
+expression_, a kind of [type expression](#type-expression). For example, in
+`T:! Hashable`, `T` is the binding (a symbolic binding in this case), and
+`Hashable` is the binding type expression.
+
+## Types and `type`
+
+A _type_ is a value of type `type`. Conversely, `type` is the type of all types.
+
+Expressions in type position, for example a [binding type expression](#bindings)
+or the return type of a function, are implicitly cast to type `type`. This means
+that it is legal to put a value that is not a type where a type is expected, as
+long as it has an implicit conversion to `type` that may be performed at compile
+time.
+
+## Facet type
+
+A _facet type_ is a [type](#types-and-type) whose values are some subset of the
+values of `type`, determined by a set of [type constraints](#type-constraints):
+
+-   [Interfaces](#interface) and [named constraints](#named-constraints) are
+    facet types whose constraints are that the interface or named constraint is
+    satisfied by the type.
+-   The values produced by `&` operations between facet types and by `where`
+    expressions are facet types, whose set of constraints are determined by the
+    `&` or `where` expression.
+-   `type` is a facet type whose set of constraints is empty.
+
+A facet type is the type used when declaring some type parameter. It foremost
+determines which types are legal arguments for that type parameter. For template
+parameters, that is all a facet type does. For checked parameters, it also
+determines the API that is available in the body of the definition of the
+[generic function, class, or other entity](#generic-means-compile-time-parameterized).
+
+## Facet
+
+A _facet_ is a value of a [facet type](#facet-type). For example,
+`i32 as Hashable` is a facet, and `Hashable` is a facet type. Note that all
+types are facets, since [`type`](#types-and-type) is considered a facet type.
+Not all facets are types, though: `i32 as Hashable` is of type `Hashable` not
+`type`, so it is a facet that is not a type. However, in places where a type is
+expected, for example in a [binding type expression](#bindings) or after the
+`->` in a function declaration, there is an automatic implicit conversion to
+`type`. This means that a facet may be used in those positions. For example, the
+facet `i32 as Hashable` will implicitly convert to `(i32 as Hashable) as type`,
+which is `i32`, in those contexts.
+
+## Type expression
+
+A _type expression_ is an expression that is being used as a type. In some
```

Comment:
I continue to be uncomfortable defining "type expression" to be a property of the context, not a property of the expression itself. Would this work instead?

```suggestion
A _type expression_ is an expression that can be used as a type. In some
```